### DISCUSSION TITLE
Pytorch is not installed on Raspberry pi 5.

### DISCUSSION THREAD

**Discussion by FAr-arch** [community/community#132638] [Programming Help]
[test_code.txt](https://github.com/user-attachments/files/16282334/test_code.txt)
### Select Topic Area

Question

### Body

Hello.
There was a problem with installing pytorch while using Raspberry pi 5.

After object recognition through image processing using Raspberry pi 5, we try to get the center coordinates of the object through object recognition and lift the object through step motor control.
Picking up items will be controlled through Bluetooth communication with Arduino.

For this, we are using yolo v5, open cv, pytorch, RPi.GPIO.

The problem is that Pytorch can only be installed in a virtual environment, so the code is executed in a virtual environment
RPi.GPIO does not work even though it is installed.
The error message is as follows.

RuntimeError : Cannot determine SOC peripheral base address

The code does not work because pytorch is not installed outside the virtual environment.
The error message is as follows.

ModuleNotFoundError : No module named 'torch'

How can I solve the problem?

Attached is the full text of my code.

import json
from time import sleep
import RPi.GPIO as gpio
import cv2
import torch
import time

# Motor pin configuration
z_direction_pin = 20  # Z-axis direction control pin number
z_pulse_pin = 21      # Z-axis pulse control pin number
cw_direction = 0      # Clockwise direction value

x_direction_pin = 6   # X-axis direction control pin number
x_pulse_pin = 5       # X-axis pulse control pin number

y_direction_pin = 24  # Y-axis direction control pin number
y_pulse_pin = 23      # Y-axis pulse control pin number

# GPIO mode and pin initialization
gpio.setmode(gpio.BCM)  # Set BCM mode (pin numbering system)
gpio.setup(z_direction_pin, gpio.OUT)  # Set Z-axis direction pin as output
gpio.setup(z_pulse_pin, gpio.OUT)      # Set Z-axis pulse pin as output
gpio.setup(x_direction_pin, gpio.OUT)  # Set X-axis direction pin as output
gpio.setup(x_pulse_pin, gpio.OUT)      # Set X-axis pulse pin as output
gpio.setup(y_direction_pin, gpio.OUT)  # Set Y-axis direction pin as output
gpio.setup(y_pulse_pin, gpio.OUT)      # Set Y-axis pulse pin as output

# Initial direction setting (all motors set to clockwise)
gpio.output(z_direction_pin, cw_direction)
gpio.output(x_direction_pin, cw_direction)
gpio.output(y_direction_pin, cw_direction)

def rotate_motor(pulse_pin, steps):
    """
    Function to rotate the motor by the specified number of steps
    :param pulse_pin: GPIO pin number to send pulse signals
    :param steps: Number of steps to rotate
    """
    for _ in range(steps):
        gpio.output(pulse_pin, gpio.HIGH)  # Set pulse signal to HIGH (start pulse)
        sleep(0.001)                       # Wait for 1 millisecond (maintain HIGH state)
        gpio.output(pulse_pin, gpio.LOW)   # Set pulse signal to LOW (end pulse)
        sleep(0.0005)                      # Wait for 0.5 milliseconds (maintain LOW state)
    print("Move to B sector complete")  # Print motor operation completion message
    process_video()                     # Call video processing function after motor operation

def process_video():
    """
    Function for video processing
    Object detection code using a given YOLOv5 model
    """
    # Load YOLOv5 model
    model = torch.hub.load('ultralytics/yolov5', 'custom', path='best2.pt', force_reload=True)

    # Start camera capture
    cap = cv2.VideoCapture(0)  # 0 represents the default camera device

    # Record the last output time
    last_time = time.time()
    fps_time = time.time()

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        # Object detection
        results = model(frame)

        # Extract necessary information from the results
        detections = results.xyxy[0].cpu().numpy()  # Convert to numpy array

        if detections.size > 0:
            # Find the object with the highest accuracy
            best_detection = max(detections, key=lambda d: d[4])

            # Calculate the central coordinates
            x_center = (best_detection[0] + best_detection[2]) / 2
            y_center = (best_detection[1] + best_detection[3]) / 2

            # Current time
            current_time = time.time()

            # Print central coordinates every 3 seconds
            if current_time - last_time >= 3:
                print(f"Center coordinates: ({x_center}, {y_center})")
                last_time = current_time

        # Calculate fps
        current_time = time.time()
        fps = 1 / (current_time - fps_time)
        fps_time = current_time

        # Add fps text to frame
        cv2.putText(frame, f"FPS: {fps:.2f}", (10, 30), cv2.FONT_HERSHEY_COMPLEX, 1, (0, 255, 0), 2, cv2.LINE_AA)

        # Render detected results on the frame
        rendered_frame = results.render()[0]
        cv2.imshow('YOLOv5 Detection', rendered_frame)

        # Exit loop when 'q' key is pressed
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    # Release capture and close window
    cap.release()
    cv2.destroyAllWindows()

class Box:
    def __init__(self, length, width, height):
        self.length = length  # Length of the box
        self.width = width    # Width of the box
        self.height = height  # Height of the box
        self.items = []       # List of items loaded in the box

    def can_fit(self, item, position):
        ix, iy, iz = item      # Dimensions of the item (length, width, height)
        px, py, pz = position  # Position to place the item (x, y, z coordinates)

        # Check if the item does not exceed the boundaries of the box
        if (px + ix <= self.length and
            py + iy <= self.width and
            pz + iz <= self.height):
            for other_item in self.items:
                opx, opy, opz, oix, oiy, oiz = other_item
                # Check if the item does not collide with other items
                if not (px + ix <= opx or opx + oix <= px or
                        py + iy <= opy or opy + oiy <= py or
                        pz + iz <= opz or opz + oiz <= pz):
                    return False
            return True
        return False

    def place_item(self, item):
        # Iterate through all positions in the box to check if the item can be placed
        for px in range(self.length):
            for py in range(self.width):
                for pz in range(self.height):
                    if self.can_fit(item, (px, py, pz)):
                        self.items.append((px, py, pz, *item))  # Place the item in the box
                        return (px, py, pz)  # Return the position where the item was placed
        return None  # If the item cannot be placed

class PackingManager:
    def __init__(self):
        self.x_positions = []  # List of x item positions
        self.y_positions = []  # List of y item positions
        self.box = None        # Box object

    def pack_items(self, x_count, y_count, box_type):
        # Set the box size based on the type of box
        if box_type == 'a':
            self.box = Box(80, 160, 60)
        elif box_type == 'b':
            self.box = Box(130, 150, 100)
        else:
            raise ValueError("Invalid box type. Choose 'a' or 'b'.")

        x_item = (30, 30, 40)  # Dimensions of x item
        y_item = (40, 40, 20)  # Dimensions of y item

        # Place x items in the box
        for _ in range(x_count):
            position = self.box.place_item(x_item)
            if position:
                self.x_positions.append(position)
            else:
                print(f"Could not fit all x_items in box {box_type}.")
                break

        # Place y items in the box
        for _ in range(y_count):
            position = self.box.place_item(y_item)
            if position:
                self.y_positions.append(position)
            else:
                print(f"Could not fit all y_items in box {box_type}.")
                break

        if box_type == 'b':
            self.move_to_center()

    def move_to_center(self):
        try:
            # Print query text and get user input
            response = input("Do you want to move to the center of sector B? (yes/no) ")

            if response.lower() == "yes":
                # Rotate motor 2 clockwise by 27.5 revolutions (assuming 200 steps per revolution)
                steps_per_revolution = 200          # Number of steps per revolution
                steps = int(27.5 * steps_per_revolution)  # Calculate the number of steps for 27.5 revolutions

                print("Rotating motor 2 clockwise by 27.5 revolutions.")
                rotate_motor(x_pulse_pin, steps)    # Perform motor 2 rotation

                print("Operation completed.")
            else:
                print("Operation terminated.")
        except KeyboardInterrupt:
            print("Terminated by user.")
        finally:
            gpio.cleanup()  # Clean up GPIO pins

    def get_packing_info(self):
        return {
            'box_dimensions': (self.box.length, self.box.width, self.box.height),
            'x_positions': self.x_positions,
            'y_positions': self.y_positions,
        }

# Main program
if __name__ == "__main__":
    x_count = int(input("Enter the number of x items: "))  # Input the number of x items
    y_count = int(input("Enter the number of y items: "))  # Input the number of y items
    box_type = input("Enter the box type (a or b): ")  # Input the type of box

    manager = PackingManager()
    manager.pack_items(x_count, y_count, box_type)

    packing_info = manager.get_packing_info()

    try:
        # Print box dimensions
        box_dimensions = packing_info['box_dimensions']
        print(f"Box dimensions: {box_dimensions[0]}x{box_dimensions[1]}x{box_dimensions[2]} (LxWxH)")

        # Print item quantities
        print(f"Number of x items: {x_count}")
        print(f"Number of y items: {y_count}")

        # Print item positions
        for i, pos in enumerate(packing_info['x_positions']):
            print(f"X item {i+1} coordinates: ({pos[0]}, {pos[1]}, {pos[2]})")

        for i, pos in enumerate(packing_info['y_positions']):
            print(f"Y item {i+1} coordinates: ({pos[0]}, {pos[1]}, {pos[2]})")
    except Exception as e:
        print(f"Error occurred: {e}")
    finally:
        gpio.cleanup()  # Clean up GPIO pins
